Add FTP DAGs

tedmiston opened the pull request:
Note: This replaces #5 with a clean diff.  I recreated the branch with a fresh branch off master replaying the relevant commits and correcting for obsoleted changes.

tedmiston commented:
@perryao Please try pulling down and see whether you experience the same issue I have with the DAG stopping after the first task.

tedmiston commented:
@perryao Please pull down with my latest commit as of a few minutes ago.  Everything is working locally for me.  Let me know if you have any comments, but I believe it's ready to merge.

One thing that bothers me is leaving print statements in here, but I already have work doing print statement to logging conversion on my refactor branch and I think it makes sense to separate that concern to that branch.  It doesn't impact the functionality here for our concerns right now.

tedmiston commented on airflow_home/dags/ftp_dags.py:
```diff
@@ -0,0 +1,97 @@
+
+# one FTP config per workflow and each customer can have zero or more workflows
+for ftp_config in ftp_configs:
+    id_ = ftp_config['_id']
```
By the way, here are some good references on the convention I mentioned if you're curious.

http://stackoverflow.com/questions/16095188/is-there-a-python-naming-convention-for-avoiding-conflicts-with-standard-module

https://shahriar.svbtle.com/underscores-in-python

perryao commented:
@tedmiston Brace yourself. Huge wall of text incoming. 

I've been doing some testing tonight and I think I've nailed down what we need to do. First, here are the requirements I'm working under:
1. A path can look like any of the following in the ftpConfig mongo document:
-  ftp/dir
- ftp/dir/
- ftp/dir/file.csv. *this allows you to specify a file with consistent names between dag runs
2. We assume that there is always one workflow, one file per directory
3. Step 1 in ftp dags is to use a sensor task (AstronomerS3WildcardSensor) to probe a path for a file upload. The return type is bool (exists or not)
4. Step 2 is to use an Operator subclass (AstronomerS3GetKeyAction) to get the full key to the file at path and xcom it's value.  The return type is String or None

Things I've confirmed we can do to achieve this are: 
1. Remove copied S3Hook and instead just subclass and override get_wildcard_key. I don't see a way around this because the current Airflow  implementation gets all the keys but only returns the first one. We need to be able to filter out all the keys ending in '/' so we ignore empty directories. So now my entire astro_s3_hook.py file looks like the following: 

```python
import re
import fnmatch
from airflow.hooks import S3Hook

class AstroS3Hook(S3Hook):

    def get_wildcard_key(self, wildcard_key, bucket_name=None, delimiter=''):
        """
        Returns a boto.s3.key.Key object matching the regular expression
        :param regex_key: the path to the key
        :type regex_key: str
        :param bucket_name: the name of the bucket
        :type bucket_name: str
        """
        if not bucket_name:
            (bucket_name, wildcard_key) = self.parse_s3_url(wildcard_key)
        bucket = self.get_bucket(bucket_name)
        prefix = re.split(r'[*]', wildcard_key, 1)[0]
        klist = self.list_keys(bucket_name, prefix=prefix, delimiter=delimiter)
        if not klist:
            return None
        key_matches = [k for k in klist if fnmatch.fnmatch(k, wildcard_key)]
        # prevent remove "directories" from returning in results as we only
        # want to match files (not an empty top-level dir)
        key_matches = [i for i in key_matches if not i.endswith('/')]
        return bucket.get_key(key_matches[0]) if key_matches else None
```
2. In order to support paths like `ftp/dir/file.csv.`, we now need to first check whether a key that  is not a directory exists in s3 for the raw path in both the AstronomerS3WildcardSensor and the AstronomerS3GetKeyAction. I've been running tests with those two classes looking like this inside astronomer_s3_key_sensor.py:
```python
class AstronomerS3WildcardKeySensor(BaseSensorOperator):
    """
    Detect a wildcard filepath in a "directory" in S3.
    """

    template_fields = ('bucket_key', 'bucket_name')

    @apply_defaults
    def __init__(self, bucket_key, bucket_name, *args, **kwargs):
        super(AstronomerS3WildcardKeySensor, self).__init__(*args, **kwargs)
        self.bucket_name = bucket_name
        self.bucket_key = bucket_key
        self.soft_fail = False

    def poke(self, context):
        hook = AstroS3Hook(s3_conn_id='S3_CONNECTION')
        # check if key exists (support for full key name)
        # need to call get_key rather than check_for_key to ignore directories
        key = hook.get_key(key=self.bucket_key, bucket_name=self.bucket_name)
        # TODO: Better python
        if key is not None and not key.name.endswith('/'):
            # is it a file or just a directory?
            return True
        # otherwise check for wildcard
        wildcard_bucket_key = os.path.join(self.bucket_key, '*')
        return hook.check_for_wildcard_key(wildcard_key=wildcard_bucket_key, bucket_name=self.bucket_name, delimiter='/')


class AstronomerS3GetKeyAction(BaseOperator):
    """
    Grab the top S3 key matching a wildcard regex.
    """

    @apply_defaults
    def __init__(self, bucket_key, bucket_name, *args, **kwargs):
        super(AstronomerS3GetKeyAction, self).__init__(*args, **kwargs)
        self.bucket_name = bucket_name
        self.bucket_key = bucket_key

    def execute(self, context):
        hook = AstroS3Hook(s3_conn_id='S3_CONNECTION')
        # check if key exists as provided
        key = hook.get_key(key=self.bucket_key, bucket_name=self.bucket_name)
        if key is not None and not key.name.endswith('/'):
            return self.format_xcom(key.name)
        else:
            # otherwise check for wildcard key
            wildcard_bucket_key = os.path.join(self.bucket_key, '*')
            wildcard_key = hook.get_wildcard_key(
                wildcard_key=wildcard_bucket_key,
                bucket_name=self.bucket_name,
                delimiter='/',
            )
            #should we instead raise an error here if None?
            return self.format_xcom(wildcard_key.name) if wildcard_key else None

    def format_xcom(self, value):
        return json.dumps({'input': {'key': value}})
```

3. Because the AstronomerS3GetKeyAction should never run unless a file actually exists, I removed the if statement you had at the bottom of execute. Wondering whether we should just  raise an error there if we get a null value. Also, I took off the xcom_push ivar. This task wouldn't make sense without xcom-ing the key it finds in s3 so we should just always assume xcom_push is true. Let me know if you were thinking of other use cases though. 
4. Because a file not landing when we expect it to is an error, I set self.soft_fail to False inside AstroS3WildcardKeySensor.

# Areas for improvement
1. I wrote a method `format_xcom` on the AstronomerS3GetKeyAction class. Probably would be better as a utility method outside of a class.
2. We should work on a better implementation of get_wildcard_key, or implement get_wildcard_keys. That way S3Hook has a way to get all keys matching a wildcard. 
3. Sometimes we use Astro, others we use Astronomer for airflow subclass prefixes. We should standardize
4. Currently all of our Operator subclasses are in the same file (airflow_home/plugins/operators/astronomer_s3_key_sensor.py). Thinking we should move each out to it's own file. 
5. I'm wondering if there's a more  pythonic way for what I'm doing here: 
```python
if key is not None and not key.name.endswith('/'):
    return self.format_xcom(key.name)
else:
```
6. astro_s3_hook.py should probably also become astronomer_s3_hook.py for consistency. Also we should move to plugins/hooks to follow the airflow convention. 
7. It'd be a hell of a lot better if we could get the exact same functionality but with only one call to s3. haven't been able to get that working for all supported path types though.

tedmiston commented:
@perryao Just wanted to reiterate that this is largely consistent with my local modifications.  I agree with most of the improvements per our discussion at the end of last week.

For implementation today, I agree `DummyOperator` is not strictly necessary.  In `ftp_dags.py`, I had it in place to have a common starting point in the case that we execute different sensors conditionally, like one for a wildcard path vs another for a non-wildcard path.

I'll circle back for review when this is all ready.

perryao commented:
@tedmiston How are things going with this?

tedmiston commented:
@perryao Totally forgot to reply here after our conversations last week, but I think we covered everything in person.  Will push my additions, changes, and revisions up for you really soon, and with that I think we can call this branch done.

perryao commented:
Thanks @tedmiston. Is that something you think can be pushed by the end of day today? First empower workflow's will be running in the early am tomorrow. If not, I can temporarily deploy to staging with what I have locally (the stuff I discussed above)

tedmiston commented:
Hey @perryao - Check out the updated version here to meet the expanded scope we discussed above / in person.

In all, I'm super happy with the progress we've made, but specifically:

- broadening support (ex. more path types than last time)
- eliminating bugs and edge cases
- hardening the sensors and operator
- decomposition and isolation of components

I came up with 9 input cases to throw at this system.  This is the same table I showed you last week but with all the blanks figured out.

This is how we handle each case after all of the revisions in this branch:

| # | s3 path format              | case                             | example                    | task status | logged | log level | 
|---|-----------------------------|----------------------------------|----------------------------|-------------|--------|-----------| 
| 1 | wildcard w/ trailing slash  | file exists                      | taylor_test/               | success     | -      | -         | 
| 2 |                             | file does not exist              | taylor_test_empty/         | failed      | yes    | WARNING   | 
| 3 |                             | directory in path does not exist | taylor_test_dne/           | failed      | yes    | WARNING   | 
| 4 | wildcard w/o trailing slash | file exists                      | taylor_test                | success     | -      | -         | 
| 5 |                             | file does not exist              | taylor_test_empty          | failed      | yes    | WARNING   | 
| 6 |                             | directory in path does not exist | taylor_test_dne            | failed      | yes    | WARNING   | 
| 7 | literal file                | file exists                      | taylor_test/dummy.csv      | success     | -      | -         | 
| 8 |                             | file does not exist              | taylor_test/dne.csv        | skipped     | yes    | WARNING   | 
| 9 |                             | directory in path does not exist | taylor_test_dne/a_file.csv | skipped     | yes    | WARNING   | 


The difference between skipped and failed task status cases are whether (1) the task had a known literal file path in advance and then that path didn't exist vs (2) a wildcard path.  (In a future branch, I'd like to adapt these cases into mocked unit tests.)

A few notes:

- I dropped the start task and everything works fine for me.  I originally had it in there as a placeholder for branching, but since we didn't end up needing that here, adios!
- Per our conversation on Wednesday, for now I've left the Airflow tasks to enter the failed state if the upstream directory doesn't exist.  We can definitely handle this more in the future if it's not clear enough to users.
- As far as naming, I dropped "astro-" everywhere I found it with re `astro[^n]`
- Since all of our components are custom, I dropped "astronomer-" in most places too as it was starting to feel redundant and probably better practice to just describe the functionality if we intend to upstream, so I've done that as much as possible.

😴

tedmiston commented:
Btw, Circle is showing me "1 failing check" on "docker build" but it seems unrelated to this branch.  Any idea about that?
